daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala

```diff
@@ -878,80 +889,174 @@ private[lf] final class Compiler(
-                SEVar(2)
+    ChoiceDefRef(tmplId, cname) ->
+      validate(
+        closureConvert(
+          Map.empty,
+          withEnv { _ =>
+            env = env.incrPos /* <actors> */
+            val selfBinderPos = env.position
+            env = env.incrPos /* <cid> */
+            val choiceArgumentPos = env.position
+            env = env.incrPos // <choice argument>
+            env = env.incrPos /* <token> */
+            // <template argument>
+            env = env.addExprVar(tmpl.param)
+            val signatories = compile(tmpl.signatories)
+            val observers = compile(tmpl.observers)
+            // now allow access to the choice argument
+            env = env.addExprVar(choice.argBinder._1, choiceArgumentPos)
+            val controllers = compile(choice.controllers)
+            val mbKey: SExpr = tmpl.key match {
+              case None => SEValue.None
+              case Some(k) => SEApp(SEBuiltin(SBSome), Array(translateKeyWithMaintainers(k)))
             }
-          )
-        },
-      ),
-    )
+            env = env.incrPos // beginExercise's ()
+            // allow access to the self contract id
+            env = env.addExprVar(choice.selfBinder, selfBinderPos)
+            val update = compile(choice.update)
+
+            withLabel(
+              s"exercise @${tmplId.qualifiedName} ${cname}",
+              SEAbs(4) {
+                SELet(
+                  // stack: <actors> <cid> <choiceArg> <token>
+                  // <tmplArg> =
+                  SBUFetch(tmplId)(SEVar(3) /* <cid> */, SEVar(1) /* <token> */ ),
+                  // stack: <actors> <cid> <choiceArg> <token> <tmplArg>
+                  // _ =
+                  SBUBeginExercise(tmplId, choice.name, choice.consuming, byKey = false)(
+                    SEVar(3), /* <choiceArg> */
+                    SEVar(4), /* <cid> */
+                    SEVar(5), /* <actors> */
+                    signatories,
+                    observers,
+                    controllers,
+                    mbKey,
+                    SEVar(2) /* <token> */
+                  ),
+                  // stack: <actors> <cid> <choiceArg> <token> <tmplArg> ()
+                  // <ret value> =
+                  SEApp(update, Array(SEVar(3) /* <token> */ )),
+                  // stack: <actors> <cid> <choiceArg> <token> <tmplArg> () <ret value>
+                  // _ =
+                  SBUEndExercise(tmplId)(
+                    SEVar(4), /* <token> */
+                    SEVar(1) /* <retValue> */
+                  )
+                ) in
+                  // stack: <actors> <cid> <choiceArg> <token> <tmplArg> () <ret value> ()
+                  SEVar(2) /* <retValue> */
+              }
+            )
+          },
+        ),
+      )
+
+  /** Compile a choice into a top-level function for exercising that choice */
+  private[this] def compileChoiceByKey(
```

Comment:
There seems to be a lot of duplication between the definition of `ChoiceDefRef` and `ChoiceByKeyDefRef`. I’m not entirely clear why we need `ChoiceByKeyDefRef` at all. What’s stopping us from compiling an `exerciseByKey` into `FetchByKeyDefRef` and `ChoiceDefRef` and removing `ChoiceByKeyDefRef`?

Reply:
The issue is that the "by cid" operations (create, exercise, fetch) have to compute the contract key from the template argument (we need the key in the transaction even if the node was built by "by cid" ). On the other hand "by key" operation don't have to do, since it is given.

I  do not see easy way to factorize as @cocreature suggested without recompute the contract key for each "by key" operations.  Note that before my PR, contract key was recomputed for `ExerciseByKey`, but not for `FetchByKey`.

We can of course, make a bit of factorization in the Scala code (I would let that for the following PR), but I think it is tricky to do it at the LF level.

Reply:
Makes sense, I’m fine with leaving it for a separate PR.